### DISCUSSION TITLE
Allowing += to be used with value from hash

### DISCUSSION THREAD

**Discussion by mitchgrout** [mesonbuild/meson#12427] [General]
Consider the following fragment:

```meson
cfg = { 'sources': [], 'includes': [] }
cfg['sources'] += files('foo', 'bar', 'baz')
cfg['includes'] += include_directories('inc')
```

Evaluating this will provide the following error

```
meson.build:2:0: ERROR: Plusassignment target must be an id.
cfg['sources'] += files('foo', 'bar', 'baz')
^
```

Is this intentional, or an oversight in the grammar/interpreter? It provides a nice way to represent data clumps (e.g. variables that do not make sense without each other), and is otherwise expressible with the more awkward

```meson
cfg = { 'sources': cfg['sources'] + files('foo', 'bar', 'baz'), 'includes': cfg['includes'] }
```

My suspicion is that the grammar specifies that `+=` is only allowed on variable-ish things, and `var[subscript]` is not recognized as variable-ish. Could this be changed?

**Comment by mitchgrout** [comment]
Addendum, this is also currently not possible:

```meson
cfg['extra'] = 42
```

with a similar message:

```
meson.build:2:0: ERROR: Assignment target must be an id.
cfg['extra'] = 42
^
```

**Comment by QuLogic** [comment]
[Dictionaries are immutable](https://mesonbuild.com/Syntax.html#dictionaries), so the assignment in your reply can't work. [Arrays are also immutable](https://mesonbuild.com/Syntax.html#arrays), and adding to them only creates a new object. So these two things together means `+=` also will not work with dictionaries, even if the values are arrays.

**Reply by mitchgrout** [reply]
What's the rationale for immutability? Does it simplify analysis of the build configuration, reduce mistakes, or something similar? In my view it limits me from more clearly expressing intent, and forces the introduction of temporary variables that clutter the configuration and make reasoning more difficult.

**Reply by xclaesse** [reply]
Mutability makes everything more complicated. If we allowed mutability of arrays, does it mean in this code `foo.c` gets compiled in my app?

```meson
sources = files(...)
executable('app', sources)
sources.append('foo.c')
```

That's opening a can of worms and we don't want the Meson language to become a fully fledged language.

About your code snippet, why are you using a dictionary at all?
```meson
sources = []
includes = []
sources += files('foo', 'bar', 'baz')
includes += include_directories('inc')
```

If you want those variable to be specific to some target, most people would just add a prefix to variable names `mylib_sources = []`.

**Reply by xclaesse** [reply]
Note that `+=` on dict/array is not mutability, it creates a new object, unlike e.g. python language.
```meson
a = [1]
b = a
a += [2]
message(a, b)
```
 That prints `[1, 2] [1]`. If that was python code it would print `[1, 2] [1, 2]`.

**Reply by xclaesse** [reply]
Note that you can do `cfg += {'sources': cfg['sources'] + files('foo', 'bar', 'baz')}`.

**Reply by mitchgrout** [reply]
I can understand a rationale for mutability there; it's a lot easier to treat structures as value types rather than reference types, and makes understanding simpler. I would then argue that there would be no reason as to why the following could not be defined:

```meson
cfg = { 'sources': [] }
cfg['sources'] += files('foo')
# is equivalent to
cfg = { 'sources': cfg['sources'] + files('foo') }
```

as it preserves the immutability in the same way as

```meson
sources = []
sources += files('foo')
# is equivalent to 
sources = sources + files('foo')
```

**Reply by mitchgrout** [reply]
While I also accept that manually expanding it would work in this case, it does not scale particularly nicely. Consider the following:

```meson
cfgs = { 
  'module-1': { 'sources': [], 'includes': [] },
  ...
  'module-N': { 'sources': [], 'includes': [] },
}

# Append a source to 'module-1'
# Proposed way:
cfgs['module-1']['sources'] += files(...)

# Current way:
cfgs  += { 'module-1': { 'sources': cfgs['module-1']['sources'] + files(...), 'includes': cfgs['module-1']['includes' } }
```

I would argue that a little syntactic sugar would go a long way here. This is not a contrived example either; I currently need to structure my project using a hash similar to the above due to needing to provide a large set of arguments to my build targets that would otherwise be unnecessarily duplicated. I also need to tweak source/include lists depending on build configuration that is otherwise not easily expressible.

**Reply by eli-schwartz** [reply]
Are there places where you could convert those arguments into `declare_dependency()` holders (yes those can handle sources too, and even depend on other declared dependencies)? You'd end up with a somewhat nested internal structure but it should in theory all work out in the end.

**Reply by mitchgrout** [reply]
The project structure is pretty convoluted for legacy reasons. The gist of it is:

1. When building my target, I provide all sources and include paths to an `executable()` call; there are no intermediate `lib` structures. I can do this with a `dep` without issue, and is indeed how I manage this portion
2. When building for tests, I instead need to build each module as a `library()` so that I can test it against test-double libraries for the rest of the system
    a. This might also involve extracting a single object from the library under test, which would need `.extract_objects(exact-path)`, which is not recoverable using a `dep` alone
 
I can see how using a utility `dep` would make my life simpler here, but unfortunately it cannot fit my entire niche.
